pkg/podmounter/mountoptions/mount_options.go

```diff
@@ -0,0 +1,130 @@
+// Package mountoptions provides utilities for passing mount options between
+// Mountpoint and CSI Driver Node Pods running in the same node.
+package mountoptions
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"syscall"
+)
+
+// An Options struct represents mount options to use while invoking Mountpoint.
+type Options struct {
+	// Fd will be passed over Unix socket using `SCM_RIGHTS`, not as part of the serialized JSON.
+	Fd         int      `json:"-"`
+	BucketName string   `json:"bucketName"`
+	Args       []string `json:"args"`
+	Env        []string `json:"env"`
+}
+
+// Send sends given mount `options` to given `sockPath` to be received by `Recv` function on the other end.
+func Send(ctx context.Context, sockPath string, options Options) error {
+	message, err := json.Marshal(&options)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message to send %s: %w", sockPath, err)
+	}
+
+	conn, err := net.Dial("unix", sockPath)
+	if err != nil {
+		return fmt.Errorf("failed to dial to unix socket %s: %w", sockPath, err)
+	}
+	defer conn.Close()
+
+	unixConn := conn.(*net.UnixConn)
+
+	unixRights := syscall.UnixRights(options.Fd)
+	messageN, unixRightsN, err := unixConn.WriteMsgUnix(message, unixRights, nil)
+	if err != nil {
+		return fmt.Errorf("failed to write to unix socket %s: %w", sockPath, err)
+	}
+	if len(message) != messageN || len(unixRights) != unixRightsN {
+		return fmt.Errorf("partial write to unix socket %s: message: size %d - written %d, unix rights: size %d - written %d",
+			sockPath, len(message), messageN, len(unixRights), unixRightsN)
+	}
+
+	return nil
+}
+
+var (
+	messageRecvSize = 1024
+	// We only pass one file descriptor and its 32 bits
+	unixRightsRecvSize = syscall.CmsgSpace(4)
+)
+
+// Recv receives passed mount options via `Send` function through given `sockPath`.
+func Recv(ctx context.Context, sockPath string) (Options, error) {
+	l, err := net.Listen("unix", sockPath)
```

Comment:
I've not looked through the protocol for this - is it documented anywhere?

Reply:
Do you mean the first parameter to `net.Listen`? If so, see https://pkg.go.dev/net#Listen

Reply:
I was more meaning the protocol for send/receive, but I can now see it's just json serialising some structure, which seems fine